implementation/BeforeInstallPromptEvent.js

```diff
@@ -0,0 +1,154 @@
+    internal.didPrompt = true;
+    const promptOutcome = await showInstallPrompt(element);
+    internal.userChoiceResolver(promptOutcome);
+  }
+
+  /**
+   * Implementation of BeforeInstallPromptEvent.
+   *
+   */
+  class BeforeInstallPromptEvent extends Event {
+
+    constructor(typeArg, eventInit) {
+      // WebIDL Guard. Not in spec, as it's all handled by WebIDL.
+      if (arguments.length === 0) {
+        throw new TypeError("Not enough arguments. Expected at least 1.");
+      }
+      const initType = typeof eventInit;
+      if (arguments.length === 2 && initType !== "undefined" && initType !== "object") {
+        throw new TypeError("Value can't be converted to a dictionary.");
+      }
+      super(typeArg, Object.assign({ cancelable: true }, eventInit));
+
+      if (eventInit && typeof eventInit.userChoice !== "undefined" && !AppBannerPromptOutcome.has(String(eventInit.userChoice))) {
+        const msg = `The provided value '${eventInit.userChoice}' is not a valid` +
+          "enum value of type AppBannerPromptOutcome.";
+        throw new TypeError(msg);
+      }
+      // End WebIDL guard.
+      const internal = {
+        didPrompt: false,
+        userChoice: null,
+        userChoiceResolver: null, // Implicit in spec
+      };
+
+      internal.userChoice = new Promise((resolve) => {
+        if (eventInit && "userChoice" in eventInit) {
+          return resolve(eventInit.userChoice);
+        }
+        internal.userChoiceResolver = resolve;
+      });
+      internalSlots.set(this, internal);
+    }
+
+    prompt() {
+      if (this.isTrusted === false) {
+        const msg = "Untrusted events can't call prompt().";
+        throw new DOMException(msg, "NotAllowedError");
+      }
+
+      if (this.defaultPrevented === false) {
+        const msg = ".prompt() needs to be called after .preventDefault()";
+        throw new DOMException(msg, "InvalidStateError");
+      }
+
+      if (internalSlots.get(this).didPrompt) {
+        const msg = ".prompt() can only be successfully called once.";
+        throw new DOMException(msg, "InvalidStateError");
+      }
+      requestInstallPrompt(this);
+    }
+
+    get userChoice() {
+      return internalSlots.get(this).userChoice;
+    }
+  }
+
+  async function notifyBeforeInstallPrompt(element) {
+    if (document.readyState === "complete") {
+      await trackReadyState();
+    }
+    if (installProcesses.length) {
+      return;
+    }
+    const event = new BeforeInstallPromptEvent("beforeinstallprompt");
+    window.dispatchEvent(event);
+    if (!event.defaultPrevented) {
+      await requestInstallPrompt(event, element);
+    }
+  }
+
+  function trackReadyState() {
+    return new Promise((resolve) => {
+      document.addEventListener("readystatechange", () => {
+        if (document.readyState === "complete") {
+          resolve();
+        }
+      });
+    });
+  }
+
+  async function showInstallPrompt(button) {
```

Comment:
the functon calling this called the argument element, why not just call it button there as well